song/song_server.py

```diff
@@ -122,3 +133,72 @@ def _send_init_to_display(self):
         self.display_client.send_message(settings.DISPLAY_INIT_ADDRESS,
                                          pickle.dumps(self.song_machine.parser.categories, protocol=2)
                                          )
+
+class Tonality:
+    '''
+    idea of tonality counter combined with rules for FX chains and Sample Triggering in
+    Ableton Live
+    '''
+
+    tonality_counter = Counter()
+    FX_KEY = 'Clean'
+    ctrl_val = 0
+    last_cats = []
+    last_value = 0
+    most_common = ''
+    tonality_lock = False
+
+    category_to_chain = {
+        # every category points to a FX Chain and a ctrl value
+        'praise': ['Delay', 5],
+        'lecture': ['Vocoder', 8],
+        'insinuation': ['FreqShift', -7],
+        'dissence': ['Distortion', -5],
+        'concession': ['Delay', 2]
+    }
+    chain_controls = {
+        # every chain has a chain_value, a ccnr and a standard ctrl_value
+        'FreqShift' : [15, 10, 65],
+        'Vocoder': [32, 15, 95],
+        'Distortion': [60, 20, 25],
+        'Delay': [90, 25, 65],
+        'Clean': [115, 0, 0]
+    }
+
+    def __init__(self, categories):
+        # self.tonality_counter = {categories[i]: 0 for i in range(len(categories))}
+        self.tonality_counter = Counter(categories)
+        self.chain = self.chain_controls[self.FX_KEY]
+
+    def update_tonality(self, cat):
+        # print("tonalities: ", self.tonality_counter)
+        self.tonality_counter[cat] += 1
+        self.calculate_rack_values(cat)
+
+    def calculate_rack_values(self, cat):
+        '''
+        1. after 10 utterances, the most_common category defines the FX chain
+        2. only if all other categories have had at least had 5 updates (self.last_cats), a new FX_kEY is
+        generated by the most_common category at that moment
+        3. the self.ctrl_value is calculated as a deviation from a standard value defined in self.chain_controls
+        '''
+        #  an FX chain is selected, if more than 10 entries have occured
+        if sum(self.tonality_counter.values()) > 10:
+            # print("cat {}   locked?: {} most common: {}".format(cat, self.tonality_lock, self.most_common))
+            if not self.tonality_lock:
+                self.most_common = [x for x,_ in self.tonality_counter.most_common(1)]
+                self.most_common = self.most_common[0]
```

Comment:
Why don't you just do `self.most_common = self.tonality_counter.most_common(1)[0][0]`?